packages/s3-store/index.ts

```diff
@@ -588,4 +612,91 @@ export class S3Store extends DataStore {
 
     this.clearCache(id)
   }
+
+  protected getExpirationDate(created_at: string) {
+    const date = new Date(created_at)
+
+    return new Date(date.getTime() + this.getExpiration())
+  }
+
+  getExpiration(): number {
+    return this.expirationPeriodInMilliseconds
+  }
+
+  async deleteExpired(): Promise<number> {
+    let keyMarker: string | undefined = undefined
+    let uploadIdMarker: string | undefined = undefined
+    let isTruncated = true
+    let deleted = 0
+
+    while (isTruncated) {
+      const listResponse: AWS.ListMultipartUploadsCommandOutput =
+        await this.client.listMultipartUploads({
+          Bucket: this.bucket,
+          KeyMarker: keyMarker,
+          UploadIdMarker: uploadIdMarker,
+        })
+
+      const expiredUploads =
+        listResponse.Uploads?.filter((multiPartUpload) => {
+          const initiatedDate = multiPartUpload.Initiated
+          return (
+            initiatedDate &&
+            this.getExpirationDate(initiatedDate.toISOString()).getTime() <
+              new Date().getTime()
+          )
+        }) || []
+
+      const objectsToDelete = expiredUploads.reduce((all, expiredUpload) => {
+        all.push(
+          {
+            key: expiredUpload.Key + '.info',
+          },
+          {
+            key: expiredUpload.Key + '.part',
+          }
+        )
+        return all
+      }, [] as {key: string}[])
+
+      const deletions: Promise<AWS.DeleteObjectsCommandOutput>[] = []
+
+      // Batch delete 1000 items at a time
```

Comment:
Isn't a 1000 concurrent connection going to be rough over a HTTP/1.1 connection, which is AFAIK what S3 still uses?

Reply:
No, we send one request with at most 1000 objects as per S3 limitation.
In case there are 1M objects expired we would send 1000 requests with each 1000 objects. 

However as far as i know the AWS SDK ships with a default connection pool of 50, so in theory should never exceed [50 requests simultaneously](https://github.com/smithy-lang/smithy-typescript/blob/main/packages/node-http-handler/src/node-http-handler.ts#L77) at any point, in order to increase or decrease this amount, you will need to pass a custom requestHandler:

```ts
const s3 = new S3Client({
  // Use a custom request handler so that we can adjust the HTTPS Agent and
  // socket behavior.
  requestHandler: new NodeHttpHandler({
    httpsAgent: new Agent({
      maxSockets: 500,

      // keepAlive is a default from AWS SDK. We want to preserve this for
      // performance reasons.
      keepAlive: true,
      keepAliveMsecs: 1000,
    }),
    socketTimeout: 5000,
  }),
});
```

Reply:
I read over the code a bit too quickly, this makes sense. Thanks.